prov/efa/src/efa_cq.c

```diff
@@ -0,0 +1,437 @@
+	entry->tag = 0;
+
+	if (ibv_wc_read_wc_flags(ibv_cqx) & IBV_WC_WITH_IMM)
+		entry->data = ibv_wc_read_imm_data(ibv_cqx);
+}
+
+/**
+ * @brief handle the situation that a TX/RX operation encountered error
+ *
+ * This function does the following to handle error:
+ *
+ * 1. write an error cq entry for the operation, if writing
+ *    CQ error entry failed, it will write eq entry.
+ *
+ * 2. increase error counter.
+ *
+ * 3. print warning message with self and peer's raw address
+ *
+ * @param[in]	base_ep     efa_base_ep
+ * @param[in]	ibv_cq_ex   extended ibv cq
+ * @param[in]	util_cq     tx/rx cq to write error to
+ * @param[in]	err         positive libfabric error code
+ * @param[in]	prov_errno  positive EFA provider specific error code
+ */
+static void efa_cq_handle_error(struct efa_base_ep *base_ep,
+				struct ibv_cq_ex *ibv_cq_ex,
+				struct util_cq *util_cq, int err,
+				int prov_errno)
+{
+	struct fi_cq_err_entry err_entry;
+	fi_addr_t addr;
+	int write_cq_err;
+
+	memset(&err_entry, 0, sizeof(err_entry));
+
+	err_entry.err = err;
+	err_entry.prov_errno = prov_errno;
+	err_entry.op_context = (void *)(uintptr_t)ibv_cq_ex->wr_id;
+	err_entry.flags = efa_cq_opcode_to_fi_flags(ibv_wc_read_opcode(ibv_cq_ex));
+	err_entry.buf = NULL;
+	err_entry.data = 0;
+	err_entry.tag = 0;
+
+	if (ibv_wc_read_wc_flags(ibv_cq_ex) & IBV_WC_WITH_IMM)
+		err_entry.data = ibv_wc_read_imm_data(ibv_cq_ex);
+
+	// TODO: get correct peer addr for tx
+	addr = efa_av_reverse_lookup(base_ep->av,
+				     ibv_wc_read_slid(ibv_cq_ex),
+				     ibv_wc_read_src_qp(ibv_cq_ex));
+
+	if (OFI_UNLIKELY(efa_write_error_msg(base_ep, addr, prov_errno,
+					     &err_entry.err_data,
+					     &err_entry.err_data_size)))
+		err_entry.err_data_size = 0;
+
+	EFA_WARN(FI_LOG_CQ, "err: %d, message: %s (%d)\n",
+		err_entry.err,
+		err_entry.err_data
+			? (const char *) err_entry.err_data
+			: efa_strerror(err_entry.prov_errno),
+		err_entry.prov_errno);
+
+	efa_show_help(err_entry.prov_errno);
+
+	efa_cntr_report_error(&base_ep->util_ep, err_entry.flags);
+	write_cq_err = ofi_cq_write_error(util_cq, &err_entry);
+	if (write_cq_err) {
+		EFA_WARN(FI_LOG_CQ,
+			"Error writing error cq entry when handling TX error\n");
+		efa_base_ep_write_eq_error(base_ep, err, prov_errno);
+	}
+}
+
+/**
+ * @brief handle the event that a request has been completed
+ *
+ * @param[in]		base_ep     efa_base_ep
+ * @param[in]		ibv_cq_ex   extended ibv cq
+ * @param[in]		cq_entry    fi_cq_tagged_entry
+ * @param[in]		is_tx       is this a TX completion
+ */
+static void efa_cq_handle_completion(struct efa_base_ep *base_ep,
+				     struct ibv_cq_ex *ibv_cq_ex,
+				     struct fi_cq_tagged_entry *cq_entry,
+				     bool is_tx)
+{
+	struct util_cq *cq = is_tx ? base_ep->util_ep.tx_cq : base_ep->util_ep.rx_cq;
+	fi_addr_t src_addr;
+	int ret = 0;
+
+	/* NULL wr_id means no completion */
+	if (!ibv_cq_ex->wr_id)
+		return;
+
+	if (!is_tx)
+		src_addr = efa_av_reverse_lookup(base_ep->av,
```

Comment:
we only need src_addr when FI_SOURCE

Reply:
Sorry about that, but now I think leaving tx/rx_handle_completion makes the code cleaner as you won't have too much `if` inside them